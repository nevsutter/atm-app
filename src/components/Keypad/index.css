.keypad {
  display: grid;
  grid-template-columns: 3fr 1fr;
  column-gap: 0.5rem;
  margin-bottom: 1rem;
}

.numericButtons, .actionButtons {
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-template-rows: repeat(4, 1fr);
  column-gap: 0.5rem;
  row-gap: 0.5rem;
}

.actionButtons {
  grid-template-columns: 1fr;
}

.numericButtons button, .actionButtons button {
  font-size: 1.2rem;
  border: none;
  border-radius: 0.2rem;
  background: #ddd;
}

.numeral0 {
  grid-column-start: 1; 
  grid-column-end: 4; 
}

button.actionClear {
  background: #dd0;
}

button.actionCancel {
  background: #b00;
}

button.actionEnter {
  background: #0d0;
  grid-row-start: 3; 
  grid-row-end: 5; 
}
