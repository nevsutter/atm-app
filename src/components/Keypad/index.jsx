/**
 * Keypad Component
 * Generates numerical and action buttons
 * Calls respective actions from props
 */
import React from 'react';
import './index.css';

function ActionButton({action, onClick}) {
  return (
    <button 
      className={`action${action}`} 
      onClick={onClick}
    >
      {action}
    </button>
  );
}

function NumericalButton({numeral, onClick}) {
  return (
    <button 
      className={`numeral${numeral}`}
      onClick={onClick}
    >
      {numeral}
    </button>
  );
}

export default class Keypad extends React.Component {
  render() {
    const numerals = [1,2,3,4,5,6,7,8,9,0];

    const numericButtons = numerals.map(numeral => {
      return (
        <NumericalButton 
          key={numeral}
          numeral={numeral}
          onClick={() => this.props.onNumberClick(numeral)}
        />
      );
    });

    return (
      <div className="keypad">
        <div className="numericButtons">{numericButtons}</div>
        <div className="actionButtons">
          <ActionButton
            action="Clear"
            onClick={this.props.onClearClick}
          />
          <ActionButton
            action="Cancel"
            onClick={this.props.onCancelClick}
          />
          <ActionButton
            action="Enter"
            onClick={this.props.onEnterClick}
          />
        </div>
      </div>
    );
  }
}