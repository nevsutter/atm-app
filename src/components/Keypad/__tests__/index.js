import Keypad from '../index';
import Enzyme,{mount} from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

Enzyme.configure({ adapter: new Adapter() });

let keypadWrapper;

beforeEach(() => {
  keypadWrapper = mount(<Keypad
    onNumberClick={(i) => {}}
    onClearClick={() => {}}
    onCancelClick={() => {}}
    onEnterClick={() => {}}
  />);
});

afterEach(() => {
  jest.resetAllMocks();
});

it('should render render correctly', () => {
  expect(keypadWrapper).toMatchSnapshot()
});

it('should render an outer div and 2 children', () => {
  expect(keypadWrapper.find('div.keypad').length).toEqual(1);
  expect(keypadWrapper.find('div.keypad div.numericButtons').length).toEqual(1);
  expect(keypadWrapper.find('div.keypad div.actionButtons').length).toEqual(1);
});

it('should render 10 numerical buttons', () => {
  expect(keypadWrapper.find('div.numericButtons button').length).toEqual(10);

  expect(keypadWrapper.find('div.numericButtons button.numeral1').length).toEqual(1);
  expect(keypadWrapper.find('div.numericButtons button.numeral2').length).toEqual(1);
  expect(keypadWrapper.find('div.numericButtons button.numeral3').length).toEqual(1);
  expect(keypadWrapper.find('div.numericButtons button.numeral4').length).toEqual(1);
  expect(keypadWrapper.find('div.numericButtons button.numeral5').length).toEqual(1);
  expect(keypadWrapper.find('div.numericButtons button.numeral6').length).toEqual(1);
  expect(keypadWrapper.find('div.numericButtons button.numeral7').length).toEqual(1);
  expect(keypadWrapper.find('div.numericButtons button.numeral8').length).toEqual(1);
  expect(keypadWrapper.find('div.numericButtons button.numeral9').length).toEqual(1);
  expect(keypadWrapper.find('div.numericButtons button.numeral0').length).toEqual(1);
});

it('should render 3 action buttons', () => {
  expect(keypadWrapper.find('div.actionButtons button').length).toEqual(3);

  expect(keypadWrapper.find('div.actionButtons button.actionClear').length).toEqual(1);
  expect(keypadWrapper.find('div.actionButtons button.actionCancel').length).toEqual(1);
  expect(keypadWrapper.find('div.actionButtons button.actionEnter').length).toEqual(1);
});

it('should call correct handler for numerical buttons', () => {
  const myFunc = jest.fn();

  keypadWrapper.setProps({
    onNumberClick: myFunc,
  });

  keypadWrapper.find('button.numeral1').simulate('click');
  keypadWrapper.find('button.numeral2').simulate('click');
  keypadWrapper.find('button.numeral3').simulate('click');
  keypadWrapper.find('button.numeral4').simulate('click');
  keypadWrapper.find('button.numeral5').simulate('click');
  keypadWrapper.find('button.numeral6').simulate('click');
  keypadWrapper.find('button.numeral7').simulate('click');
  keypadWrapper.find('button.numeral8').simulate('click');
  keypadWrapper.find('button.numeral9').simulate('click');
  keypadWrapper.find('button.numeral0').simulate('click');

  expect(myFunc).toBeCalledTimes(10);
  
  expect(myFunc).toBeCalledWith(1);
  expect(myFunc).toBeCalledWith(2);
  expect(myFunc).toBeCalledWith(3);
  expect(myFunc).toBeCalledWith(4);
  expect(myFunc).toBeCalledWith(5);
  expect(myFunc).toBeCalledWith(6);
  expect(myFunc).toBeCalledWith(7);
  expect(myFunc).toBeCalledWith(8);
  expect(myFunc).toBeCalledWith(9);
  expect(myFunc).toBeCalledWith(0);
});

it('should call correct handler for clear button', () => {
  const myFunc = jest.fn();

  keypadWrapper.setProps({
    onClearClick: myFunc,
  });

  keypadWrapper.find('button.actionClear').simulate('click');

  expect(myFunc).toBeCalled();
});

it('should call correct handler for cancel button', () => {
  const myFunc = jest.fn();

  keypadWrapper.setProps({
    onCancelClick: myFunc,
  });

  keypadWrapper.find('button.actionCancel').simulate('click');

  expect(myFunc).toBeCalled();
});

it('should call correct handler for enter button', () => {
  const myFunc = jest.fn();

  keypadWrapper.setProps({
    onEnterClick: myFunc,
  });

  keypadWrapper.find('button.actionEnter').simulate('click');

  expect(myFunc).toBeCalled();
});
