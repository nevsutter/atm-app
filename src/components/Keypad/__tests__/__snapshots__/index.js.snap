// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render render correctly 1`] = `ReactWrapper {}`;
